scalding-commons/src/main/scala/com/twitter/scalding/commons/source/LzoGenericScheme.scala

```diff
@@ -0,0 +1,118 @@
+  val inj: Injection[Externalizer[_], String] = {
+    import com.twitter.bijection.Inversion.attemptWhen
+    import com.twitter.bijection.codec.Base64
+
+    implicit val baseInj = JavaSerializationInjection[Externalizer[_]]
+
+    implicit val unwrap: Injection[GZippedBase64String, String] =
+      new AbstractInjection[GZippedBase64String, String] {
+        override def apply(gzbs: GZippedBase64String) = gzbs.str
+        override def invert(str: String) = attemptWhen(str)(Base64.isBase64)(GZippedBase64String(_))
+      }
+
+    Injection.connect[Externalizer[_], Array[Byte], GZippedBase64String, String]
+  }
+}
+
+private[source] object ConfigBinaryConverterProvider {
+  val ProviderConfKey = "com.twitter.scalding.lzo.converter.provider"
+}
+
+/**
+ * Provides BinaryConverter serialized in JobConf.
+ */
+private[source] class ConfigBinaryConverterProvider[M] extends BinaryConverterProvider[M] {
+  import ConfigBinaryConverterProvider._
+  override def getConverter(conf: Configuration): BinaryConverter[M] = {
+    val data = conf.get(ProviderConfKey)
+    require(data != null, s"No data in field $ProviderConfKey")
+
+    val extern: Externalizer[_] = ExternalizerSerializer.inj.invert(data).get
+    extern.get.asInstanceOf[BinaryConverter[M]]
+  }
+}
+
+/**
+ * Generic scheme for data stored as lzo-compressed protobuf messages.
+ * Serialization is performed using the supplied BinaryConverter.
+ */
+class LzoGenericScheme[M: ClassTag](@transient conv: BinaryConverter[M]) extends LzoBinaryScheme[M, GenericWritable[M]] {
+
+  override protected def prepareBinaryWritable(): GenericWritable[M] =
+    new GenericWritable(conv)
+
+  override def sourceConfInit(fp: FlowProcess[JobConf],
+    tap: Tap[JobConf, RecordReader[_, _], OutputCollector[_, _]],
+    conf: JobConf): Unit = {
+
+    val extern = Externalizer(conv)
+    try {
+      ExternalizerSerializer.inj.invert(ExternalizerSerializer.inj(extern)).get
+    } catch {
+      case e: Exception => throw new RuntimeException("Unable to roundtrip the BinaryConverter in the Externalizer.", e)
+    }
+
+    conf.set(ConfigBinaryConverterProvider.ProviderConfKey, ExternalizerSerializer.inj(extern))
+
+    MultiInputFormat.setClassConf(conv.getClass, conf)
+    MultiInputFormat.setGenericConverterClassConf(classOf[ConfigBinaryConverterProvider[_]], conf)
+
+    DelegateCombineFileInputFormat.setDelegateInputFormat(conf, classOf[MultiInputFormat[_]])
+  }
+
+  override def sinkConfInit(fp: FlowProcess[JobConf],
+    tap: Tap[JobConf, RecordReader[_, _], OutputCollector[_, _]],
+    conf: JobConf): Unit = {
+    val extern = Externalizer(conv)
+    try {
+      ExternalizerSerializer.inj.invert(ExternalizerSerializer.inj(extern)).get
+    } catch {
+      case e: Exception => throw new RuntimeException("Unable to roundtrip the BinaryConverter in the Externalizer.", e)
+    }
+
+    LzoGenericBlockOutputFormat.setClassConf(implicitly[ClassTag[M]].runtimeClass, conf)
```

Comment:
would 'implicitly[ClassTag[M]].runtimeClass' be the thrift class? Can we use that in sourceConfInit too? This will help with stats and logging (where we would see the actual thrift class)